### ISSUE TITLE
Allow install of mods on offline machines

### ISSUE THREAD

**Issue by Mikaz-fr** [ramensoftware/windhawk#259]
With https://github.com/ramensoftware/windhawk/issues/234 an offline installer was provided, allowing offline machines to use Windhawk.
But I can't find any information or ways to install mods offline.
The mods repository do not allow to download the mods and the program do not allow to load from disk it seems.

Is there any way to install mods to an offline machine other than creating new ones locally on the machine ?

**Comment by m417z** [comment]
> Is there any way to install mods to an offline machine other than creating new ones locally on the machine ?

No, currently that's the easiest way. I'm not sure how often such offline mod installation is needed for users, but it's a valid suggestion, I'll consider it for one of the future versions.

**Comment by m417z** [comment]
For now, it's probably easier to go to the changelog, click on the desired version date to open the code on GitHub, and copy/download it from there. It has more convenient copy controls, and it's also allows to install any version.

**Comment by chocmake** [comment]
Was curious if this was feasible for an entirely offline system of mine and tested with the [Vertical Taskbar](https://windhawk.net/mods/taskbar-vertical) mod, by copying the source code to a file then manually transferring to the system and pasting in the template file for 'Create new mod'.

However it errored when compiling, with `Failed to parse settings: not a valid YAML array`. Is copy-pasting mod source all that's necessary? As it reads from the OP like this is possible.

(I also tested in an online system using the same method in case there was some dependency that would download but the same error occurred.)

**Comment by Mikaz-fr** [comment]
> However it errored when compiling, with `Failed to parse settings: not a valid YAML array`.

You most likely didn’t copy the full code properly. As indicated above it’s easier to copy from the GitHub code. If you copy from the Windhawk mod page you must disable the button `Collapse Readme and Settings` on the page so the full settings are copied.

**Comment by chocmake** [comment]
Thanks, that was it.

(Though after compiling it seems the mod has no affect on the latest W11 24H2, at least virtualized (and using the portable Windhawk) but that would be a separate issue)

**Comment by 4R05T** [comment]
Is it possible to have an option for the mods to be either zip files or windhawk pkgs (like rainmeter pkgs for example) so one can download the pkg and be able to flexibly install it on any pc, offline or not

**Comment by Darthagnon** [comment]
Is there a way to use Windhawk fully offline? I'm in the UK, and recently [Imgur has been blocking all the screenshots of mods on all my machines](https://en.shiftdelete.net/imgur-blocked-in-the-uk/), making even the official online experience lacklustre. I'm worried for the future availability or if I want to enable mods while travelling. 

I have cloned all the mod repositories to my offline machine, but there is no way to point Windhawk at the full repositories and have it work... 

> List of all repos for offline use:
> - Offline Windhawk installer from https://github.com/ramensoftware/windhawk/releases
> - https://github.com/ramensoftware/windhawk-mod-symbol-cache
> - https://github.com/ramensoftware/windhawk-symbol-helper
> - https://github.com/ramensoftware/resource-redirect-themes
> - Most screenshots are hosted externally on Imgur, which is blocked in some countries, rather than in the Github repos... (Manual download of Imgur screenshots needed)
> - Windows debug symbols also needed, see below...

I have them all downloaded, is there any way to just use it offline?

--------------------------------

Windows debug symbols can be downloaded for offline use (and I block Windows update to avoid breakages, should be less of a problem now that Win10 is EOL):

- [PDB-Downloader program](https://github.com/rajkumar-rangaraj/PDB-Downloader)
- [PowerShell script, based off PDB-Downloader](https://github.com/FranciscoNabas/PowerShellPublic/blob/main/Get-PdbSymbol.ps1)
- https://web.archive.org/web/20160730104526/https://developer.microsoft.com/en-us/windows/hardware/download-symbols
- https://web.archive.org/web/20170930104816/https://developer.microsoft.com/en-us/windows/hardware/download-symbols

Further reading:
- https://github.com/valinet/ExplorerPatcher/wiki/Symbols
- https://www.bleepingcomputer.com/news/microsoft/microsoft-is-no-longer-providing-offline-msi-symbol-packages/
- https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-symbols
- https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/using-a-manifest-file-with-symchk
- https://stackoverflow.com/questions/50179282/how-to-load-all-windows-symbols-from-server-starting-with-w10-version-1803-bu

**Comment by hllshiro** [comment]
We have a lot of machines running on our LAN, and it would be great to have an easy way to install mods offline.

**Comment by Darthagnon** [comment]
A manual way to install symbols offline:

### m417z said:
> The error is still ERROR_INTERNET_CANNOT_CONNECT. Are you able to download the file via the browser?
https://msdl.microsoft.com/download/symbols/Taskbar.pdb/0A9CC91C50EAEEBD9DB4F952F1C3233D1/Taskbar.pdb
> 
> You can copy it manually to `C:\ProgramData\Windhawk\Engine\Symbols`:
> 
> - Open `C:\ProgramData\Windhawk\Engine\Symbols\`
> - Create the folders `Taskbar.pdb\0A9CC91C50EAEEBD9DB4F952F1C3233D1\`
> - Put the downloaded file inside the newly downloaded folder: `C:\ProgramData\Windhawk\Engine\Symbols\Taskbar.pdb\0A9CC91C50EAEEBD9DB4F952F1C3233D1\`
> - Rename it to Taskbar.pdb
> - Disable and re-enable the mod
> - Look at the logs, and do the same for any other missing symbol files.
> 
> All this should be done automatically, but because you're using an old Windows build and you have a problem with your internet connection, this is the easiest way to make it work.

The above can be done manually for a collection of previously manually downloaded symbols for your specific build of Windows. It is likely very manual and time consuming, but doable.
